src/Search.svelte

```diff
@@ -1,42 +1,196 @@
-<script lang="ts">
+<script lang="ts" context="module">
+  import type { Host } from "@app/api";
+  import type { ProjectInfo } from "@app/project";
+
+  import { ethers } from "ethers";
+
+  import * as utils from "@app/utils";
+  import { Profile } from "@app/profile";
+  import { Project } from "@app/project";
+
+  export interface ProjectsAndProfiles {
+    projects: { info: ProjectInfo; seed: Host }[];
+    profiles: Profile[];
+  }
+
+  type SearchResult =
+    | { type: "nothing" }
+    | { type: "error"; message: string }
+    | { type: "singleProfile"; id: string }
+    | { type: "singleProject"; seedHost: string; id: string }
+    | { type: "projectsAndProfiles"; projectsAndProfiles: ProjectsAndProfiles };
+
+  async function searchProjectsAndProfiles(
+    query: string,
+    config: Config,
+  ): Promise<SearchResult> {
+    try {
+      // The query is a plain Ethereum address.
+      if (ethers.utils.isAddress(query)) {
+        return { type: "singleProfile", id: query };
+      }
+
+      const projectOnSeeds = Object.keys(config.seeds.pinned).map(seed => ({
+        nameOrUrn: query,
+        seed,
+      }));
+
+      // The query is a radicle project URN.
+      if (utils.isRadicleId(query)) {
+        const projects = await Project.getMulti(projectOnSeeds);
+
+        if (projects.length === 1) {
+          return {
+            type: "singleProject",
+            seedHost: projects[0].seed.host,
+            id: query,
+          };
+        } else {
+          return {
+            type: "projectsAndProfiles",
+            projectsAndProfiles: { projects, profiles: [] },
+          };
+        }
+      }
+
+      // The query is either a project or a profile name.
+      const normalizedQuery = query.toLowerCase();
+      const projectsAndProfiles: ProjectsAndProfiles = {
+        projects: [],
+        profiles: [],
+      };
+
+      try {
+        const projects = await Project.getMulti(projectOnSeeds);
+        projectsAndProfiles.projects.push(...projects);
+      } catch {
+        // TODO: collect errors and forward to user.
+      }
+
+      try {
+        let params: string[];
+        if (normalizedQuery.includes(".")) {
```

Comment:
I suppose you could use here the `isEnsName` util function.
```suggestion
        if (isEnsName(normalizedQuery, config)) {
```

Reply:
51853d2100558473d1f6bd62176291c6fd507185